### PR TITLE
fix sdpa math accuracy issue when scale is negative

### LINKED ISSUE #105190: [torch.compile] `sfdp_replacement_1` returns NaN for the model
### 🐛 Describe the bug

A model will return `NaN` after `sfdp_replacement_1` optimization in `torch.compile`

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()

    def forward(self, x):
        v1 = torch.matmul(x, x.transpose(-1, -2))
        v2 = v1 / -0.0001
        v3 = v2.softmax(dim=-1)
        v4 = torch.matmul(v3, x)
        return v4

func = Model().to('cpu')

x = torch.randn(1, 3, 64, 64)

test_inputs = [x]

with torch.no_grad():
    func.train(False)
    jit_func = torch.compile(func)

    res1 = func(x) # without jit
    print(res1)
    # tensor([[[[-0.4655, -1.1350, -0.9982,  ..., -1.7541, -1.5332,  0.8824],
    #           [ 1.3429, -1.4892, -1.2500,  ..., -0.5561, -1.4006,  0.3309],
    #           [ 1.3429, -1.4892, -1.2500,  ..., -0.5561, -1.4006,  0.3309],
    #           ...,
    #           [ 0.2517,  0.0069,  1.0887,  ..., -0.5190,  0.1727, -0.5762],
    #           [ 1.7818,  1.1335, -1.0080,  ...,  2.5257,  0.5339, -1.4125],
    #           [ 0.5597,  0.2945,  1.1403,  ..., -0.7962, -1.5412, -0.3925]],
    # 
    #          [[-0.6230, -1.9249, -0.3286,  ..., -1.0716, -0.0655, -1.7286],
    #           [ 1.3485, -1.3512,  0.3916,  ..., -0.3752,  0.1259,  1.5306],
    #           [ 0.0446,  0.7670, -1.8892,  ...,  0.2005,  1.7442, -2.2904],
    #           ...,
    #           [-0.1974,  1.1985, -0.0783,  ...,  1.0809,  0.5481,  1.8275],
    #           [-0.2876,  0.7077, -2.5406,  ..., -1.6377, -1.0338, -1.2810],
    #           [-0.1646, -1.6991,  0.3815,  ..., -0.3185, -0.0520, -0.8371]],
    # 
    #          [[ 0.0433,  0.4435,  1.2171,  ..., -0.1094,  1.5639,  1.6918],
    #           [-1.0094, -1.3880, -0.1377,  ..., -0.4373,  0.0737, -1.1173],
    #           [-1.8482,  0.3627,  0.5636,  ...,  0.6504, -0.9623, -0.8336],
    #           ...,
    #           [ 0.3773, -0.6312, -2.1172,  ..., -2.3102, -0.7575, -0.7120],
    #           [-0.7257,  1.8288, -1.2434,  ...,  0.9904, -3.1187,  1.2762],
    #           [ 1.0475, -1.4160,  0.3584,  ..., -0.5997,  0.4141, -0.3347]]]])

    res2 = jit_func(x)
    print(res2)
    # tensor([[[[nan, nan, nan,  ..., nan, nan, nan],)
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230713+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+3c400e7818
[pip3] torch==2.1.0.dev20230713+cu118
[pip3] torchaudio==2.1.0.dev20230713+cu118
[pip3] torchvision==0.16.0.dev20230713+cu118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+3c400e7818          pypi_0    pypi
[conda] torch                     2.1.0.dev20230713+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230713+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230713+cu118          pypi_0    pypi

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105202

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp
Comment (jgong5) [inline aten/src/ATen/native/transformers/attention.cpp:639]: Are you still able to invoke `sdp::calculate_scale` without reimplementing its logic here?

File: aten/src/ATen/native/transformers/attention.cpp
Comment (jgong5) [inline aten/src/ATen/native/transformers/attention.cpp:667]: I guess you can make the input of `at::matmul` negative to avoid an extra post-op here?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/transformers/attention.cpp
Comment (XiaobingSuper) [reply aten/src/ATen/native/transformers/attention.cpp:667]: but SymFloat doesn't have neg method: https://github.com/pytorch/pytorch/blob/main/c10/core/SymFloat.h/

File: aten/src/ATen/native/transformers/attention.cpp
Comment (XiaobingSuper) [reply aten/src/ATen/native/transformers/attention.cpp]: updated.

**Review by jgong5 — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
@ngimel @malfet, could you help review this PR?

File: aten/src/ATen/native/transformers/attention.cpp
Comment (lezcano) [inline aten/src/ATen/native/transformers/attention.cpp:634]: Why doing all this here rather than teach `sdp::calculate_scale` how to handle negative scales?

**Review by lezcano — APPROVED** [review]
Fair enough

File: aten/src/ATen/native/transformers/attention.cpp
Comment (lezcano) [reply aten/src/ATen/native/transformers/attention.cpp:634]: Now, that being said, do we want the CPU path to deviate from the CUDA path?

**Review by lezcano — COMMENTED** [review]
Actually, sorry for the noise, but before merging this, could we have a test that compares CPU and CUDA. We wouldn't want to have an implementation where one deviates from the other.

**Comment by XiaobingSuper** [comment]
> Actually, sorry for the noise, but before merging this, could we have a test that compares CPU and CUDA. We wouldn't want to have an implementation where one deviates from the other.

Yes, I can add a test to cover it.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (XiaobingSuper) [reply aten/src/ATen/native/transformers/attention.cpp:634]: So we need to make sure the ```_scaled_dot_product_attention_math``` will get the same value with the cuda fused ```_scaled_dot_product_attention``` path, is it right?

Comment (lezcano) [reply aten/src/ATen/native/transformers/attention.cpp:634]: Yeah, I believe so (modulo precision errors, of course)

**Review by drisspg — CHANGES_REQUESTED** [review]
This logic should be handled in the calculate_scale argument and needs to be documented.

File: test/test_transformers.py
Comment (XiaobingSuper) [inline test/test_transformers.py:1511]: @lezcano @drisspg ， I think this test case cover that CPU and GPU path get same result for math path when scale<0 wchih we always compare sdpb result with a ref path.

**Review by lezcano — APPROVED** [review]
Nice! Thank you!

Let's wait for @drisspg review though before merging.

**Review by drisspg — APPROVED** [review]
This looks good,  the fused kernels should handle the negative scaling correctly, wouldn't mind a small test for that but otherwise this looks good

**Pull request closed without merging**
